lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp

```diff
@@ -0,0 +1,217 @@
+#include "rclcpp/rclcpp.hpp"
+#include "cmath"
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+#include "chrono"
+#include "kdl_parser/kdl_parser.hpp"
+#include "kdl/chainfksolverpos_recursive.hpp" // for forward kinematics
+#include "kdl/chainiksolverpos_lma.hpp" // for inverse kinematics
+#include "lbr_fri_msgs/msg/lbr_position_command.hpp"
+#include "lbr_fri_msgs/msg/lbr_state.hpp"
+#include "friClientIf.h"
+
+
+using namespace std::chrono_literals; // this namespace to enable we can use 500ms to represent 500 milliseconds in our code.
+using std::placeholders::_1;
+
+lbr_fri_msgs::msg::LBRState current_robot_state; // store current robot state, including measured joint positions
+KDL::Frame initial_cartesian_position; // we want to move the end effector in sine curve in z-direction in Cartesian space, this is the initial position 
+
+class JointPositionPublisher:public rclcpp::Node
+{
+	private:
+		rclcpp::Publisher<lbr_fri_msgs::msg::LBRPositionCommand>::SharedPtr publisher_;
+        rclcpp::Subscription<lbr_fri_msgs::msg::LBRState>::SharedPtr subscriber_;
+
+        KDL::Chain chain_; // robot kinematics chain exetracted from robot URDF file
+
+        double amplitude_; // rad
+        double frequency_; // Hz
+        double sampling_time_; // sampling time for the state of robot and sending position command
+        double phase_; // initial phase
+		int count_;
+
+	private:
+        /**
+         * @function: callback function for Robot State Subscriber
+        */
+        void topic_callback(const lbr_fri_msgs::msg::LBRState& msg)
+        {
+            current_robot_state = msg;
+            
+            if(count_ < 200) // in first 2 seconds, to make 'initial_position_command' be 'measured_joint_position'
+            {
+                std::cout << "count_" << count_ << std::endl;
+                double joint_position[7];
+                for(int i = 0; i < 7; i++)
+                {
+                    joint_position[i] = current_robot_state.measured_joint_position[i]; 
+                }
+                initial_cartesian_position = computeForwardKinematics(chain_, joint_position);
+            }
+            else
+            {
+                if (current_robot_state.session_state == KUKA::FRI::COMMANDING_ACTIVE) 
+                {
+                    KDL::Frame cartesian_position_command = initial_cartesian_position;
+
+                    cartesian_position_command.p = KDL::Vector(initial_cartesian_position.p.x(), initial_cartesian_position.p.y(), initial_cartesian_position.p.z() + amplitude_ * sin(phase_));
```

Comment:
oh okay I see, this performs oscillation in Cartesian space, could you maybe add a Cartesian pose topic instead?